src/main/client/batch_read.c

```diff
@@ -0,0 +1,299 @@
+
+        // NOTE these conversions shouldn't go wrong but if they do, return
+        if (key_to_pyobject(&err, res->key, &py_key) != AEROSPIKE_OK) {
+            as_log_error("unable to convert res->key at results index: %d", i);
+            success = false;
+            break;
+        }
+
+        // Create BatchRecord instance
+        py_batch_record = PyObject_CallMethodObjArgs(
+            data->batch_records_module, data->func_name, py_key, NULL);
+        if (py_batch_record == NULL) {
+            as_log_error("unable to instance BatchRecord at results index: %d",
+                         i);
+            success = false;
+            Py_DECREF(py_key);
+            break;
+        }
+        Py_DECREF(py_key);
+
+        // Initialize BatchRecord instance
+        as_batch_result_to_BatchRecord(data->client, &err, res, py_batch_record,
+                                       data->checking_if_records_exist);
+        if (err.code != AEROSPIKE_OK) {
+            as_log_error(
+                "as_batch_result_to_BatchRecord failed at results index: %d",
+                i);
+            success = false;
+            Py_DECREF(py_batch_record);
+            break;
+        }
+
+        PyList_Append(data->py_results, py_batch_record);
+        Py_DECREF(py_batch_record);
+    }
+
+    PyGILState_Release(gstate);
+    return success;
+}
+
+PyObject *AerospikeClient_BatchRead(AerospikeClient *self, PyObject *args,
+                                    PyObject *kwds)
+{
+    PyObject *py_keys = NULL;
+    PyObject *py_bins = NULL;
+    PyObject *py_policy_batch = NULL;
+    static char *kwlist[] = {"keys", "bins", "policy", NULL};
+    if (PyArg_ParseTupleAndKeywords(args, kwds, "O|OO:batch_read", kwlist,
+                                    &py_keys, &py_bins,
+                                    &py_policy_batch) == false) {
+        return NULL;
+    }
+
+    as_error err;
+    as_error_init(&err);
+
+    // required arg so don't need to check for NULL
+    if (!PyList_Check(py_keys)) {
+        as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                        "keys should be a list of aerospike key tuples");
+        goto CLEANUP1;
+    }
+
+    as_vector tmp_keys;
+    Py_ssize_t keys_size = PyList_Size(py_keys);
+    as_vector_init(&tmp_keys, sizeof(as_key), keys_size);
+    as_vector *tmp_keys_p = &tmp_keys;
+
+    if (!self || !self->as) {
+        as_error_update(&err, AEROSPIKE_ERR_PARAM, "Invalid aerospike object");
+        goto CLEANUP2;
+    }
+
+    if (!self->is_conn_16) {
+        as_error_update(&err, AEROSPIKE_ERR_CLUSTER,
+                        "No connection to aerospike cluster");
+        goto CLEANUP2;
+    }
+
+    uint64_t processed_key_count = 0;
+    for (int i = 0; i < keys_size; i++) {
+        PyObject *py_key = PyList_GetItem(py_keys, i);
+        as_key *tmp_key = (as_key *)as_vector_get(&tmp_keys, i);
+
+        Py_XINCREF(py_key);
+        if (!PyTuple_Check(py_key)) {
```

Comment:
Hey @dwelch-spike, `py_key` shouldn't ever be NULL. The only time `PyList_GetItem()` returns NULL is when the index is out of bounds, but that shouldn't happen since we are iterating from 0 to `keys_size`, and `keys_size` was set to `PyList_Size(py_keys)` earlier in the code.